evennia/contrib/pathfinding/scripts.py

```diff
@@ -0,0 +1,96 @@
+"""
+PathfinderScript
+
+Contrib - Johnny 2018
+
+A batteries-included abstraction for in-MUD navigation.
+
+Upon initialization, the Pathfinder will quietly build a "map" (directional 
+graph) of all of your game's Rooms and Exits, even accounting for one-way 
+travel.
+
+Requires the `networkx` library, not included with Evennia, but only a
+`pip install networkx` away.
+
+---
+
+Installation:
+
+Get or create an instance of this script at server start/reload.
+
+To do so, modify `yourgame/server/conf/at_server_startstop.py`:
+
+Add this import to the top of the file:
+
+    from evennia.contrib.pathfinding.scripts import PathfinderScript
+    
+Add this line to the beginning or end of your `at_server_start()` function, but
+above the word 'pass' if it's there:
+
+    # This will preload graph creation before the game is live
+    pathfinder = PathfinderScript.spawn()
+    
+That's it! It is now running and will update itself every 30 minutes.
+
+To use it:
+
+    from evennia.contrib.pathfinding.scripts import PathfinderScript
+    pfind = PathfinderScript.spawn()
+    
+    pfind.map.get_directions(room_obj1, room_obj2)
+    >> ['down', 'east', 'down', 'north', 'down']
+    
+"""
+
+from evennia import DefaultScript
+from evennia.contrib.pathfinding.pathfinder import Pathfinder
+from evennia.utils import create, logger
+
+class PathfinderScript(DefaultScript):
+    """Generates and maintains a global Pathfinder instance."""
+    
+    @property
+    def map(self):
+        """
+        Returns a reference to the stored Pathfinder.
+        """
+        return self.pathfinder
+        
+    @classmethod
+    def spawn(cls, **kwargs):
+        """
+        Gets the first existing script instance, or creates new
+        
+        Kwargs:
+            any (any): Forwarded to create.create_script
+            
+        Returns:
+            obj (PathfinderScript): PathfinderScript
+        """
+        if kwargs:
+            obj = cls.objects.get(**kwargs)
+        else:
+            obj = cls.objects.first()
+            
+        kwargs['obj'] = None
+        if not obj: obj = create.create_script(cls, **kwargs)
+        return obj
+    
+    def at_script_creation(self):
+        self.key = "pathfinder"
+        self.desc = "Maintains a global Pathfinder instance."
+        self.interval = 60 * 30 # every 30 minutes
```

Comment:
This sounds like it would be suitable to set during creation rather than having it hard-coded. For a production game that is not into a phase of heavy building, running this every 24h seems more than enough.

Reply:
That's what I tried to facilitate but I don't fully understand script construction. Any kwargs (like interval) passed to the spawn() function get forwarded to create.create_script(), and I was under the assumption constructor kwargs took precedence over anything hardcoded. I don't see any other way to pass data to at_script_creation. Advice plz? :)

Reply:
No, that sounds right, I think I just missed that spawn forwards to `create_script`, sorry.